### PR TITLE
[CSM] Add send_chat_message and run_server_chatcommand API functions

### PR DESCRIPTION
**Pull request by paly2**
This is basically a copy of https://github.com/minetest/minetest/pull/5452, I don't understand why this PR has been closed. For convenience, and to avoid adding legacy stuff in case sending server chat commands gets a bit more complicated, I added a `minetest.run_server_chatcommand` alias.

Possibly needed changes:
* Adding type checks for `minetest.run_server_chatcommand` params

### CODE REVIEW

**Comment by nerzhul** [comment]
client side rate limiting should be added

**Comment by paly2** [comment]
Done

**Comment by nerzhul** [comment]
the queue is a quite good idea, but i'm not very fan. If you want a queue, please limit the queue size and refuse if there are too many messages

**Comment by paly2** [comment]
Done. The queue size limit defaults to 50 and can be disabled by setting to -1 (0 disables queueing). It can be cleared using the `minetest.clear_out_chat_queue` function or the `.clear_chat_queue` client chatcommand.

**Comment by red-001** [comment]
is the chatcommand really need?

**Comment by paly2** [comment]
> is the chatcommand really need?

I think so, especially when you paste too much text in the console (same problem as on IRC, most clients have a `/flushq` command)

**Comment by paly2** [comment]
Are there remaining changes to do or is it ready for merge?

File: src/client.cpp [hunk 1]
```diff
@@ -1155,13 +1165,49 @@ void Client::sendInventoryAction(InventoryAction *a)
...
+
+	return true;
+}
+
 void Client::sendChatMessage(const std::wstring &message)
 {
-	NetworkPacket pkt(TOSERVER_CHAT_MESSAGE, 2 + message.size() * sizeof(u16));
+	const s16 max_queue_size = g_settings->getS16("max_out_chat_queue_size");
+	if (canSendChatMessage()) {
+		u32 now = time(NULL);
```
Comment (nerzhul) [inline src/client.cpp]: seems your forgot to remove this code which is now in canSendChatMessage :)

File: src/client.cpp [hunk 2]
```diff
@@ -1155,13 +1165,49 @@ void Client::sendInventoryAction(InventoryAction *a)
...
+		m_out_chat_queue.push(message);
+	} else {
+		infostream << "Coulnd't queue chat message because maximum out chat queue size ("
+				<< max_queue_size << ") is reached." << std::endl;
+	}
+}
+
+void Client::clearOutChatQueue()
+{
+	m_out_chat_queue = std::queue<std::wstring>();
```
Comment (nerzhul) [inline src/client.cpp]: not the good way to empty a std container. Please use clear method on the queue

**Comment by nerzhul** [comment]
can you fix the remaining point ?

**Comment by paly2** [comment]
@nerzhul I can't fix it, see my replies

File: src/client.cpp (hunk 2 above)
Comment (nerzhul) [inline src/client.cpp]: okay !

File: po/fr/minetest.po [hunk 3]
```diff
@@ -19,6 +19,14 @@ msgstr ""
...
 "X-Generator: Weblate 2.10-dev\n"
 
+#: builtin/client/chatcommands.lua
+msgid "Clear the out chat queue"
+msgstr "Nettoie la file d'attente de discussion en sortie"
+
+#: builtin/client/chatcommands.lua
+msgid "The out chat queue is now empty"
+msgstr "La file d'attente de discussion en sortie a été nettoyée"
+
```
Comment (SmallJoker) [inline po/fr/minetest.po]: Will be done by weblate automatically.

File: src/client.cpp [hunk 4]
```diff
@@ -1155,13 +1165,49 @@ void Client::sendInventoryAction(InventoryAction *a)
 	Send(&pkt);
 }
 
+bool Client::canSendChatMessage() const
+{
+	u32 now = time(NULL);
+	float time_passed = now - m_last_chat_message_sent;
+
+	float virt_chat_message_allowance = m_chat_message_allowance + time_passed * (CLIENT_CHAT_MESSAGE_LIMIT_PER_10S / 8.0f);
```
Comment (SmallJoker) [inline src/client.cpp]: Line too long

File: src/client.cpp [hunk 5]
```diff
@@ -1155,13 +1165,49 @@ void Client::sendInventoryAction(InventoryAction *a)
...
 {
-	NetworkPacket pkt(TOSERVER_CHAT_MESSAGE, 2 + message.size() * sizeof(u16));
+	const s16 max_queue_size = g_settings->getS16("max_out_chat_queue_size");
+	if (canSendChatMessage()) {
+		u32 now = time(NULL);
+		float time_passed = now - m_last_chat_message_sent;
+		m_last_chat_message_sent = time(NULL);
 
-	pkt << message;
+		m_chat_message_allowance += time_passed * (CLIENT_CHAT_MESSAGE_LIMIT_PER_10S / 8.0f);
```
Comment (SmallJoker) [inline src/client.cpp]: Why `/ 8.0f`?

File: src/client.cpp [hunk 6]
```diff
@@ -1155,13 +1165,49 @@ void Client::sendInventoryAction(InventoryAction *a)
...
+
+		NetworkPacket pkt(TOSERVER_CHAT_MESSAGE, 2 + message.size() * sizeof(u16));
+
+		pkt << message;
+
+		Send(&pkt);
+	} else if (m_out_chat_queue.size() < (u16) max_queue_size || max_queue_size == -1) {
+		m_out_chat_queue.push(message);
+	} else {
+		infostream << "Coulnd't queue chat message because maximum out chat queue size ("
```
Comment (SmallJoker) [inline src/client.cpp]: *Couldn't   or Could not

File: builtin/settingtypes.txt [hunk 7]
```diff
@@ -312,6 +312,9 @@ serverlist_url (Serverlist URL) string servers.minetest.net
 #    File in client/serverlist/ that contains your favorite servers displayed in the Multiplayer Tab.
 serverlist_file (Serverlist file) string favoriteservers.txt
 
+#    Maximum size of the out chat queue. 0 to disable queueing and -1 to make the queue size illimited
+max_out_chat_queue_size (Maximum size of the out chat queue) int 50
```
Comment (SmallJoker) [inline builtin/settingtypes.txt]: Missing in `minetest.conf.example`

File: src/client.cpp (hunk 5 above)
Comment (paly2) [inline src/client.cpp]: I'm wondering, too. @nerzhul wrote this algorithm on server-side (https://github.com/minetest/minetest/blob/master/src/remoteplayer.cpp#L199)

File: src/script/lua_api/l_client.cpp [hunk 8]
```diff
@@ -75,6 +75,25 @@ int ModApiClient::l_display_chat_message(lua_State *L)
...
+	lua_pushboolean(L, true);
+	return 1;
+}
+
+// clear_out_chat_queue()
+int ModApiClient::l_clear_out_chat_queue(lua_State *L)
+{
+	getClient(L)->clearOutChatQueue();
+	lua_pushboolean(L, true);
+	return 1;
```
Comment (SmallJoker) [inline src/script/lua_api/l_client.cpp]: A function that always returns the same value doesn't need a return value.

File: builtin/settingtypes.txt (hunk 7 above)
Comment (paly2) [inline builtin/settingtypes.txt]: For the default queue size, what value would be better?

File: src/script/lua_api/l_client.cpp (hunk 8 above)
Comment (paly2) [inline src/script/lua_api/l_client.cpp]: Every CSM functions do so.

File: po/fr/minetest.po (hunk 3 above)
Comment (paly2) [inline po/fr/minetest.po]: Okay, should I remove it?

File: builtin/settingtypes.txt (hunk 7 above)
Comment (SmallJoker) [inline builtin/settingtypes.txt]: I think 20 are already enough for the general purpose.

File: src/script/lua_api/l_client.cpp (hunk 8 above)
Comment (SmallJoker) [inline src/script/lua_api/l_client.cpp]: Not inconsistent. See `send_respawn`, `sound_stop` and `take_screenshot`.

Comment (paly2) [inline src/script/lua_api/l_client.cpp]: Oh well, so the file is _already_ inconsistent.

File: po/fr/minetest.po (hunk 3 above)
Comment (SmallJoker) [inline po/fr/minetest.po]: It will be overwritten by the translation from weblate sooner or later anyway, so I think you can keep it.

**Comment by paly2** [comment]
@SmallJoker Everything's now ixed

**Pull request merged**